src/ast/rewriter/bit_blaster/bit_blaster_adder.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "ast/ast.h"
+#include "ast/rewriter/bool_rewriter.h"
+#include "util/rational.h"
+
+class bit_blaster_adder {
+public:
+    typedef rational numeral;
+
+    bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, numeral const & value = numeral::zero());
+    bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, expr * const * bits);
+
+    bit_blaster_adder(bool_rewriter & rewriter, expr_ref_vector & value):
+        bit_blaster_adder(rewriter, value.size(), value.data()) {}
+
+    bit_blaster_adder(bit_blaster_adder const & other):
+        bit_blaster_adder(other.m_rewriter, other.size()) {
+        *this += other;
+    }
+
+    bit_blaster_adder(bit_blaster_adder &&) noexcept = default;
+
+    // The *_bits functions all return the final carry bit out of the addition;
+    // just ignore it if you don't need it.
+
+    // Return the sum of the known-constant inputs to this adder.
+    bool constant_bits(numeral & value) const {
+        value = m_constant % power(size());
+        return m_constant.get_bit(size());
+    }
+
+    // Return the sum of the non-constant inputs to this adder.
+    expr_ref variable_bits(expr_ref_vector & out_bits) const;
+
+    // Return the sum of all inputs to this adder.
+    expr_ref total_bits(expr_ref_vector & out_bits) const;
+
+    unsigned size() const {
+        return m_variable.size();
+    }
+
+    ast_manager & m() const {
+        return m_rewriter.m();
+    }
+
+    bit_blaster_adder & operator+=(bit_blaster_adder const & other) {
+        SASSERT(size() == other.size());
+        return add_shifted(other, 0);
+    }
+
+    bit_blaster_adder & add_bit(unsigned idx, bool bit) {
+        SASSERT(idx < size());
+        if (bit)
+            m_constant += power(idx);
+        return *this;
+    }
+
+    bit_blaster_adder & add_bit(unsigned idx, expr * bit) {
+        SASSERT(idx < size());
+        if (m().is_true(bit))
+            m_constant += power(idx);
+        else if (!m().is_false(bit))
+            m_variable[idx].push_back(bit);
+        return *this;
+    }
+
+    bit_blaster_adder & add_bit(unsigned idx, expr_ref & bit) {
+        return add_bit(idx, bit.get());
+    }
+
+    bit_blaster_adder & add_shifted(bit_blaster_adder const & other, unsigned shift);
+    bit_blaster_adder & add_shifted(unsigned sz, expr * const * bits, unsigned shift);
+
+    bit_blaster_adder & add_shifted(expr_ref_vector const & bits, unsigned shift) {
+        return add_shifted(bits.size(), bits.data(), shift);
+    }
+
+    bit_blaster_adder & add_shifted(numeral const & value, unsigned shift) {
+        m_constant += value * power(shift);
+        reduce();
+        return *this;
+    }
+
+protected:
+    bool_rewriter & m_rewriter;
```

Comment:
hopefully the m_rewriter that gets passed in is not configured with "flat" set to "true".
There was a long latent bug where the bit-blaster would convert into circuits but break all structure sharing by setting flat to true.

Reply:
I've added a call to `rewriter.set_flat(false)`. I'm not sure how to evaluate what impact that had on memory usage, but the unit tests I wrote went from reporting `:after-memory 0.13` to `:after-memory 0.02`, so I hope that's a positive sign?